### PR TITLE
Implement ju.TreeMap

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

**Comment by gzm0** [comment]
Depends on #4797.

**Comment by gzm0** [comment]
@sjrd I'm not sure whether this is enough test coverage. It does re-use RedBlackTree, so we don't really need to test the core more, but probably some "minor" forwarders are untested.

File: javalib/src/main/scala/java/util/TreeMap.scala [hunk 1]
```diff
@@ -0,0 +1,602 @@
...
+    headMap(toKey, false)
+
+  def tailMap(fromKey: K): SortedMap[K, V] =
+    tailMap(fromKey, true)
+}
+
+private object TreeMap {
+  private class ProjectedEntrySet[K, V](tree: RB.Tree[K, V],
+      lowerBound: K, lowerKind: RB.BoundKind, upperBound: K, upperKind: RB.BoundKind)(
+      implicit protected val comp: Comparator[_ >: K]) extends AbstractSet[Map.Entry[K, V]] {
```
Comment (sjrd) [inline javalib/src/main/scala/java/util/TreeMap.scala:254]:
```suggestion
      implicit protected val comp: Comparator[_ >: K])
      extends AbstractSet[Map.Entry[K, V]] {
```

File: javalib/src/main/scala/java/util/TreeMap.scala [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+      subMap(fromKey, true, toKey, false)
+
+    def headMap(toKey: K): SortedMap[K, V] =
+      headMap(toKey, false)
+
+    def tailMap(fromKey: K): SortedMap[K, V] =
+      tailMap(fromKey, true)
+
+
+    // Common implementation of pollFirstEntry() and pollLastEntry()
```
Comment (sjrd) [inline javalib/src/main/scala/java/util/TreeMap.scala:404]: Double blank line.

**Review by sjrd — CHANGES_REQUESTED** [review]
The tests look good enough to me. This looks great.
I only have very minor comments.

**Comment by sjrd** [comment]
The CI says:
```
[error] C:\projects\scala-js\test-suite\shared\src\test\scala\org\scalajs\testsuite\javalib\util\TreeMapTest.scala:68:33: inferred type arguments [K,?0] do not conform to method comparing's type parameter bounds [T,U <: Comparable[_ >: U]]
[error]     val natural = ju.Comparator.comparing((_: K).asInstanceOf[Comparable[Any]])
[error]                                 ^
[error] C:\projects\scala-js\test-suite\shared\src\test\scala\org\scalajs\testsuite\javalib\util\TreeMapTest.scala:68:62: type mismatch;
[error]  found   : java.util.function.Function[K,Comparable[Any]]
[error]  required: java.util.function.Function[_ >: T, _ <: U]
[error]     val natural = ju.Comparator.comparing((_: K).asInstanceOf[Comparable[Any]])
[error]                                                              ^
[error] two errors found
[error] (testSuite2_12 / Test / compileIncremental) Compilation failed
```

**Comment by gzm0** [comment]
Rebased and addressed review comments.

**Review by sjrd — APPROVED** [review]

**Comment by sjrd** [comment]
The following commands are failing in the CI:
```
sbt "set scalaJSLinkerConfig in testSuite.v2_12 ~= makeCompliant" testSuite2_12/test
```
On JDK 16, the JVM test suite fails:
```
sbt testSuiteJVM2_12/test testSuiteExJVM2_12/test
```

**Comment by gzm0** [comment]
Seems this is affected by: https://bugs.openjdk.org/browse/JDK-8259622?subTaskView=unresolved

**Comment by gzm0** [comment]
Updated. For ease of review, I have made two commits to be squashed.

**Review by sjrd — APPROVED** [review]
LGTM after squashing.

**Pull request merged**